cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,725 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+                            } else if (*opr.op == operator_type::NEQ) {
+                                return !equal(opr.rhs, cvs, bag);
+                            } else if (opr.op->is_slice()) {
+                                return limits(opr, bag);
+                            } else if (*opr.op == operator_type::CONTAINS) {
+                                return contains(opr.rhs->bind_and_get(bag.options), cvs, bag);
+                            } else if (*opr.op == operator_type::CONTAINS_KEY) {
+                                return contains_key(cvs, opr.rhs->bind_and_get(bag.options), bag);
+                            } else if (*opr.op == operator_type::LIKE) {
+                                return like(cvs, *opr.rhs, bag);
+                            } else if (*opr.op == operator_type::IN) {
+                                return is_one_of(cvs, *opr.rhs, bag);
+                            } else {
+                                throw exceptions::unsupported_operation_exception("Unhandled binary_operator");
+                            }
+                        },
+                        // TODO: implement.
+                        [] (const token& tok) -> bool {
+                            throw exceptions::unsupported_operation_exception("token");
+                        },
+                    }, opr.lhs);
+            },
+        }, restr);
+}
+
+} // anonymous namespace
+
+bool is_satisfied_by(
+        const expression& restr,
+        const std::vector<bytes>& partition_key, const std::vector<bytes>& clustering_key,
+        const query::result_row_view& static_row, const query::result_row_view* row,
+        const selection& selection, const query_options& options) {
+    const auto regulars = get_non_pk_values(selection, static_row, row);
+    return is_satisfied_by(
+            restr, {options, row_data_from_partition_slice{partition_key, clustering_key, regulars, selection}});
+}
+
+bool is_satisfied_by(
+        const expression& restr,
+        const schema& schema, const partition_key& key, const clustering_key_prefix& ckey, const row& cells,
+        const query_options& options, gc_clock::time_point now) {
+    return is_satisfied_by(restr, {options, row_data_from_mutation{key, ckey, cells, schema, now}});
+}
+
+std::vector<bytes_opt> first_multicolumn_bound(
+        const expression& restr, const query_options& options, statements::bound bnd) {
+    auto found = find_if(restr, [bnd] (const binary_operator& oper) {
+        return matches(oper.op, bnd) && std::holds_alternative<std::vector<column_value>>(oper.lhs);
+    });
+    if (found) {
+        auto oper = std::get<binary_operator>(*found);
+        return static_pointer_cast<tuples::value>(oper.rhs->bind(options))->get_elements();
+    } else {
+        return std::vector<bytes_opt>{};
+    }
+}
+
+value_set possible_lhs_values(const expression& expr, const query_options& options) {
```

Comment:
The explanation is above the function declaration in restriction.hh:
```c++
/// A vector of all LHS values that would satisfy an expression.  Assumes all expression's atoms have the same
/// LHS, which is either token or a single column_value.
```

> How would you represent `WHERE a = b`?

I wouldn't; it's an error to invoke this function on that expression.  This is used in calculating bounds ranges for slice fetch.

> what if func is token?

Token is statically typed, so visitors handle it separately.  A generic function should result in unbounded slice, rather than calling this.

Reply:
Ok, please expand the function name to indicate it works on a subset of the expression.

I'd also like special cases so that simple expressions like ck = ? don't work by creating an infinite range and then intersecting it with the bind variable.

Reply:
> I'd also like special cases so that simple expressions like ck = ? don't work by creating an infinite range and then intersecting it with the bind variable.

This is not necessary.  The very purpose of the `query_options` parameter is to bind and evaluate all markers in stride.

Reply:
I decided not to change the function name, as it says precisely what the function returns.  Instead, I added a column_definition parameter (so it's clear whose values are being sought) and expanded the blurb.

I also don't think we should explain cases like `WHERE a=b` that aren't supported by the current expression hierarchy nor the CQL parser.